### PR TITLE
Fix scheduler incorrect steps when gradient accumulation enabled

### PR DESCRIPTION
**Pull request by markovalexander**
Fix https://github.com/huggingface/accelerate/issues/963 issue

### CODE REVIEW

File: src/accelerate/test_utils/scripts/test_sync.py [hunk 1]
```diff
@@ -202,11 +202,13 @@ def test_gradient_accumulation_with_opt_and_scheduler(split_batches=False, dispa
...
-            sched.step()
-        else:
-            for _ in range(accelerator.num_processes):
+
+        if ((iteration + 1) % 2 == 0) or ((iteration + 1) == len(dataloader)):
+            if split_batches:
                 sched.step()
+            else:
+                for _ in range(accelerator.num_processes):
+                    sched.step()
```
Comment (sgugger) [inline src/accelerate/test_utils/scripts/test_sync.py:211]: Could you explain he logic here?

**Review by sgugger — COMMENTED** [review]
Thanks for your PR. The firs half LGTM as discussed on the issue, but I don't fully understand the changes on the test file. Could you elaborate?

File: src/accelerate/test_utils/scripts/test_sync.py (hunk 1 above)
Comment (markovalexander) [reply src/accelerate/test_utils/scripts/test_sync.py:211]: Yes, basically I added if-statement to check  that optimizer step was made, as in line 224: `did_step = (((iteration + 1) % 2) == 0) or ((iteration + 1) == len(dataloader)) `. Because AcceleratedScheduler now steps every `gradient_accumulation_steps` steps, I added same logic to the ground truth scheduler. May be I should have moved `opt.step()` under this if statement too.

**Review by sgugger — APPROVED** [review]
Thanks for the explanation! @muellerzr can you also have a look here?

**Review by muellerzr — APPROVED** [review]
Thanks! This all looks good to me as well. I'll keep an eye post-merge to make sure our multi-GPU tests all pass (as those are only ran on merges/nightly) but great work @markovalexander! Thanks! 🤗

**Comment by yjkim3** [comment]
From this commit therefore,

Do I don't need to set the steps of scheduler with calculating on my own?
Is scheduler steps internally adjusted? (or Is sheduler.step() runed every steps in regardless to gradient_accumulation_steps)

If I try to configure below,
training steps = 1000
scheduler steps = 1000
gradient_accumulation_steps = 4

then, is the scheduler.step() runed total 250 steps (1000 / 4)?

image

**Comment by muellerzr** [comment]
Hi!

> Is scheduler.step() run every steps in regardless to gradient_accumulation_steps
No it's run only on steps that actually "step", and then it is run `n` times for `n` GPUs. 

It will still run `1000` steps, we only step the scheduler when we step, as I mentioned before

**Comment by yjkim3** [comment]
> Hi!
> 
> > Is scheduler.step() run every steps in regardless to gradient_accumulation_steps
> > No it's run only on steps that actually "step", and then it is run `n` times for `n` GPUs.
> 
> It will still run `1000` steps, we only step the scheduler when we step, as I mentioned before (aka update the weights on all processes/synchronize)

Thanks for explanation :)
